### ISSUE TITLE
Crash with concurrent iterations and writes

### ISSUE THREAD

**Issue by mrunge** [redhat-service-assurance/smart-gateway#22]
```
019/06/04 07:50:29 Debug:Adding heartbeat for host host3734.
2019/06/04 07:50:29 Debug:Getting metrics for host host4462  with total plugin size 1
2019/06/04 07:50:29 Debug:Getting metrics for host host3176  with total plugin size 1
2019/06/04 07:50:29 Debug:Adding heartbeat for host host3176.
fatal error: concurrent map iteration and map write
2019/06/04 07:50:29 time %!s(int=2)

goroutine 924 [running]:
runtime.throw(0xb09d4c, 0x26)
	/usr/lib/golang/src/runtime/panic.go:608 +0x72 fp=0xc0002f0d28 sp=0xc0002f0cf8 pc=0x439232
runtime.mapiternext(0xc0002f0e68)
	/usr/lib/golang/src/runtime/map.go:790 +0x525 fp=0xc0002f0db0 sp=0xc0002f0d28 pc=0x41bad5
github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics.(*cacheHandler).Collect(0xc0001c0ab0, 0xc000d0df80)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics/metrics.go:56 +0x114 fp=0xc0002f0ed8 sp=0xc0002f0db0 pc=0x975e44
github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/registry.go:430 +0x193 fp=0xc0002f0fe0 sp=0xc0002f0ed8 pc=0x775093
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc0002f0fe8 sp=0xc0002f0fe0 pc=0x4693d1
created by github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/registry.go:522 +0xe23

goroutine 1 [select]:
github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics.StartMetrics()
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics/metrics.go:284 +0x1019
main.main()
	/go/src/github.com/redhat-service-assurance/smart-gateway/cmd/main.go:19 +0x168

goroutine 19 [syscall, 2 minutes]:
os/signal.signal_recv(0x0)
	/usr/lib/golang/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/lib/golang/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/lib/golang/src/os/signal/signal_unix.go:29 +0x41

goroutine 52 [chan receive, 2 minutes]:
github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics.StartMetrics.func2(0xc0001845a0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics/metrics.go:190 +0xe4
created by github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics.StartMetrics
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics/metrics.go:189 +0x792

goroutine 53 [runnable]:
github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil.IncomingDataCache.Put(0xc00025d500, 0x12c, 0xc000258440, 0xc000b20b70, 0x8)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil/cacheserver.go:98 +0xf9
github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil.IncomingDataCache.GetShard(0xc00025d500, 0x12c, 0xc000258440, 0xc000b20b70, 0x8, 0xc0008f2d80)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil/cacheserver.go:127 +0xd5
github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil.CacheServer.loop(0xc00025d500, 0x12c, 0xc000258440, 0xc000184660, 0x0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil/cacheserver.go:225 +0xad
created by github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil.NewCacheServer
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil/cacheserver.go:196 +0x11e

goroutine 62 [IO wait, 2 minutes]:
internal/poll.runtime_pollWait(0x7fe6414d2f00, 0x72, 0x0)
	/usr/lib/golang/src/runtime/netpoll.go:173 +0x66
internal/poll.(*pollDesc).wait(0xc000260398, 0x72, 0xc0000a8000, 0x0, 0x0)
	/usr/lib/golang/src/internal/poll/fd_poll_runtime.go:85 +0x9a
internal/poll.(*pollDesc).waitRead(0xc000260398, 0xffffffffffffff00, 0x0, 0x0)
	/usr/lib/golang/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc000260380, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/internal/poll/fd_unix.go:384 +0x1a0
net.(*netFD).accept(0xc000260380, 0x50, 0xa6ec60, 0xc0002eed01)
	/usr/lib/golang/src/net/fd_unix.go:238 +0x42
net.(*TCPListener).accept(0xc0001b62e8, 0xc0002eed90, 0xc000585810, 0x3320ceb7be7dbf39)
	/usr/lib/golang/src/net/tcpsock_posix.go:139 +0x2e
net.(*TCPListener).AcceptTCP(0xc0001b62e8, 0xc0002eedb8, 0x496ab6, 0x5cf62229)
	/usr/lib/golang/src/net/tcpsock.go:247 +0x47
net/http.tcpKeepAliveListener.Accept(0xc0001b62e8, 0xc0002eee08, 0x18, 0xc000178780, 0x6ae0d5)
	/usr/lib/golang/src/net/http/server.go:3232 +0x2f
net/http.(*Server).Serve(0xc00018b6c0, 0xb95e80, 0xc0001b62e8, 0x0, 0x0)
	/usr/lib/golang/src/net/http/server.go:2826 +0x22f
net/http.(*Server).ListenAndServe(0xc00018b6c0, 0xc00018b6c0, 0x20)
	/usr/lib/golang/src/net/http/server.go:2764 +0xb6
net/http.ListenAndServe(0xc0001be910, 0xc, 0xb90be0, 0xc00025d920, 0x1, 0xc0001be910)
	/usr/lib/golang/src/net/http/server.go:3004 +0x74
github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics.StartMetrics.func4(0xc00025d920)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics/metrics.go:236 +0x197
created by github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics.StartMetrics
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics/metrics.go:233 +0xb2a

goroutine 66 [select]:
github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp.(*AMQPServer).start(0xc000302000, 0x12f8b00)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp/receiver.go:328 +0x221
created by github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp.NewAMQPServer
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp/receiver.go:132 +0x1e6

goroutine 67 [runnable]:
syscall.Syscall(0x1, 0x2, 0xc000590000, 0x24, 0x24, 0x24, 0x0)
	/usr/lib/golang/src/syscall/asm_linux_amd64.s:18 +0x5
syscall.write(0x2, 0xc000590000, 0x24, 0x380, 0xc00016e001, 0x0, 0x0)
	/usr/lib/golang/src/syscall/zsyscall_linux_amd64.go:1005 +0x5a
syscall.Write(0x2, 0xc000590000, 0x24, 0x380, 0x9a, 0xc0002efbf8, 0x495254)
	/usr/lib/golang/src/syscall/syscall_unix.go:191 +0x49
internal/poll.(*FD).Write(0xc0000a60c0, 0xc000590000, 0x24, 0x380, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/internal/poll/fd_unix.go:268 +0x165
os.(*File).write(0xc0000a4010, 0xc000590000, 0x24, 0x380, 0x6, 0x32, 0x4)
	/usr/lib/golang/src/os/file_unix.go:266 +0x4e
os.(*File).Write(0xc0000a4010, 0xc000590000, 0x24, 0x380, 0x12f8ba0, 0x0, 0x0)
	/usr/lib/golang/src/os/file.go:145 +0x6f
log.(*Logger).Output(0xc0000b2000, 0x2, 0xc000a7c320, 0xf, 0x0, 0x0)
	/usr/lib/golang/src/log/log.go:172 +0x20d
log.Printf(0xaf69b0, 0x7, 0xc000418a20, 0x1, 0x1)
	/usr/lib/golang/src/log/log.go:295 +0x80
github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp.NewAMQPServer.func1(0xaf69b0, 0x7, 0xc000418a20, 0x1, 0x1)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp/receiver.go:127 +0x53
github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp.(*AMQPServer).start.func1(0xc000302000, 0xc0002f8180, 0x0, 0xc000300180, 0xc0002f81e0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp/receiver.go:264 +0x799
created by github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp.(*AMQPServer).start
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/amqp/receiver.go:251 +0x14d

goroutine 74 [select]:
github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton.(*Engine).Run(0xc000370090, 0x0, 0x0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton/engine.go:392 +0x468
github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/electron.(*connection).run(0xc0003145b0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/electron/connection.go:210 +0x44
created by github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/electron.NewConnection
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/electron/connection.go:202 +0x394

goroutine 7 [IO wait]:
internal/poll.runtime_pollWait(0x7fe6414d2e30, 0x72, 0xc00004ee08)
	/usr/lib/golang/src/runtime/netpoll.go:173 +0x66
internal/poll.(*pollDesc).wait(0xc000302498, 0x72, 0xffffffffffffff00, 0xb91ae0, 0x12c9950)
	/usr/lib/golang/src/internal/poll/fd_poll_runtime.go:85 +0x9a
internal/poll.(*pollDesc).waitRead(0xc000302498, 0x7fe620007e00, 0x4000, 0x4000)
	/usr/lib/golang/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc000302480, 0x7fe620007e10, 0x4000, 0x4000, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/internal/poll/fd_unix.go:169 +0x179
net.(*netFD).Read(0xc000302480, 0x7fe620007e10, 0x4000, 0x4000, 0xc00004ef18, 0x4127bb, 0xc0003a8060)
	/usr/lib/golang/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc000312040, 0x7fe620007e10, 0x4000, 0x4000, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/net/net.go:177 +0x68
github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton.(*Engine).Run.func1(0xc00002a030, 0xc0003a8060, 0xc000370090, 0xc0003a8120)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton/engine.go:336 +0xd8
created by github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton.(*Engine).Run
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton/engine.go:329 +0x1a6

goroutine 8 [chan receive]:
github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton.(*Engine).Run.func2(0xc00002a030, 0xc0003a80c0, 0xc000370090, 0xc0003a8180)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton/engine.go:352 +0x88
created by github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton.(*Engine).Run
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/qpid.apache.org/proton/engine.go:349 +0x1ec

goroutine 65 [runnable]:
github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.(*constMetric).Write(0xc000d9bf20, 0xc00093f1f0, 0xc0000ef840, 0x0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/value.go:118 +0x82
github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.processMetric(0xb938c0, 0xc000d9bf20, 0xc0000ef810, 0xc0000ef840, 0x0, 0x0, 0x0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/registry.go:550 +0x9b
github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather(0xc0000b2370, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/registry.go:476 +0x878
github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.UninstrumentedHandler.func1(0xb95480, 0xc0001b6000, 0xc00028c500)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/http.go:75 +0x47
net/http.HandlerFunc.ServeHTTP(0xb198f8, 0xb95480, 0xc0001b6000, 0xc00028c500)
	/usr/lib/golang/src/net/http/server.go:1964 +0x44
net/http.Handler.ServeHTTP-fm(0xb95480, 0xc0001b6000, 0xc00028c500)
	/usr/lib/golang/src/net/http/h2_bundle.go:5592 +0x4d
github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.InstrumentHandlerFuncWithOpts.func1(0xb95a00, 0xc000dec000, 0xc00028c500)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/http.go:289 +0x23e
net/http.HandlerFunc.ServeHTTP(0xc0001c8690, 0xb95a00, 0xc000dec000, 0xc00028c500)
	/usr/lib/golang/src/net/http/server.go:1964 +0x44
net/http.(*ServeMux).ServeHTTP(0xc00025d920, 0xb95a00, 0xc000dec000, 0xc00028c500)
	/usr/lib/golang/src/net/http/server.go:2361 +0x127
net/http.serverHandler.ServeHTTP(0xc00018b6c0, 0xb95a00, 0xc000dec000, 0xc00028c500)
	/usr/lib/golang/src/net/http/server.go:2741 +0xab
net/http.(*conn).serve(0xc000308280, 0xb96140, 0xc0001cb640)
	/usr/lib/golang/src/net/http/server.go:1847 +0x646
created by net/http.(*Server).Serve
	/usr/lib/golang/src/net/http/server.go:2851 +0x2f5

goroutine 923 [semacquire]:
sync.runtime_Semacquire(0xc000dee068)
	/usr/lib/golang/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc000dee060)
	/usr/lib/golang/src/sync/waitgroup.go:130 +0x64
github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func2(0xc000dee060, 0xc000d0df80, 0xc000a78ea0)
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/registry.go:447 +0x2b
created by github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/go/src/github.com/redhat-service-assurance/smart-gateway/vendor/github.com/prometheus/client_golang/prometheus/registry.go:446 +0x5cf

goroutine 920 [IO wait]:
internal/poll.runtime_pollWait(0x7fe6414d2d60, 0x72, 0xc000729e58)
	/usr/lib/golang/src/runtime/netpoll.go:173 +0x66
internal/poll.(*pollDesc).wait(0xc00080c118, 0x72, 0xffffffffffffff00, 0xb91ae0, 0x12c9950)
	/usr/lib/golang/src/internal/poll/fd_poll_runtime.go:85 +0x9a
internal/poll.(*pollDesc).waitRead(0xc00080c118, 0xc000364600, 0x1, 0x1)
	/usr/lib/golang/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc00080c100, 0xc000364641, 0x1, 0x1, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/internal/poll/fd_unix.go:169 +0x179
net.(*netFD).Read(0xc00080c100, 0xc000364641, 0x1, 0x1, 0xc0001cb718, 0xc000729f68, 0x47d5d5)
	/usr/lib/golang/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc00073c448, 0xc000364641, 0x1, 0x1, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/net/net.go:177 +0x68
net/http.(*connReader).backgroundRead(0xc000364630)
	/usr/lib/golang/src/net/http/server.go:676 +0x5a
created by net/http.(*connReader).startBackgroundRead
	/usr/lib/golang/src/net/http/server.go:672 +0xd2
```

**Comment by abays** [comment]
In `internal/pkg/cacheutil/cacheserver.go`, a semaphore is used within `GetHosts()`, but this does not protect access to the map that is then returned:

```
//GetHosts  Get All hosts
func (i IncomingDataCache) GetHosts() map[string]*ShardedIncomingDataCache {
	i.lock.Lock()
	defer i.lock.Unlock()
	return i.hosts
}
```

The caller would also need to lock the semaphore before `range`ing through the hosts map.  Thus, the locking mechanism in this function is useless.  This is exactly the situation encountered in `internal/pkg/metrics/metrics.go` lines 54+:

```
allHosts := c.cache.GetHosts()
debugm("Debug:Prometheus is requesting to scrape metrics...")
for key, plugin := range allHosts {    // <--- allHosts is unprotected
    ...
}
```

So if something else writes to this map while the range is iterating (or vice-versa), you get the panic.

**Comment by paramite** [comment]
@abays Thanks Andrew, your comment helped a lot!

**Comment by abays** [comment]
Just some more postmortem analysis...

You can see from the stack trace that a thread was writing to the hosts cache when the panic occurred:

```
goroutine 53 [runnable]:
github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil.IncomingDataCache.Put(0xc00025d500, 0x12c, 0xc000258440, 0xc000b20b70, 0x8)
	/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil/cacheserver.go:98 +0xf9
```

The `Put` function honors the semaphore and writes to the cache...

```
func (i IncomingDataCache) Put(key string) {
	i.lock.Lock()
	defer i.lock.Unlock()
	i.hosts[key] = NewShardedIncomingDataCache(i.maxTTL)
}
```

...but of course the read-loop I mentioned above (in my first comment) does not acknowledge the semaphore.  If that loop was iterating through the hosts cache in another thread when this thread tried to write to the hosts cache, it would explain why the panic happened.  With both the write and the read-loop threads honoring the semaphore, however, we shouldn't see this again.

**Comment by leifmadsen** [comment]
Resolved by PR #23
